src/web/demo/src/test/java/org/geoserver/web/demo/MapPreviewPageTest.java

```diff
@@ -246,20 +252,4 @@ public void testNameURLEncoding() {
             catalog.save(ft);
         }
     }
-
-    /** Test for layer group service support check */
-    @Test
-    public void testHasServiceSupport() throws Exception {
```

Comment:
GeoServer tests are slow to start due to partial application startup, moving them around in their own test class slows the build for everybody. What was the reason for the move?

Reply:
You suggested to move the hasServiceSupport method into the WMS facade class (which I did not know what that was).  However, I found that moving the method into the PreviewLayer class (and making it public) had the effect of making it available in the MapPreviewPage.java where it came from and was still needed, as well as in the CommonFormatBean and its subclasses, by virtue of it being a parameter of the abstract getFormatLink(PreviewLayer layer) method, where it is also needed.  That move eliminated the wmsVisible and wfsVisible parameters from the getFormatLink(...) call, and now the implementations of CommonFormatBean can calculate their own visibility as needed with no / minimal overhead.

So I created a test class for the PreviewLayer class, (which did not have any tests), in order to move the test method.  Given that PreviewLayer doesn't have any other tests, I would be willing to write some, if you think it's a good idea, and if you'll give me some pointers on what might be useful.  Maybe I used the wrong base class?  Could a different base class reduce the overhead of setup in this case?

Reply:
Hum... if you move to have a more unit test in the other class, please still leave the integration test here, GeoServer has no testers and no form of functional testing, so integration tests like this one are the only proof that something is actually working. Unit testing can also be added of course, but no development is complete until there are integration tests making sure there are no regressions.

That said, hasServiceSupport needs a fully started GeoServer, I don't see an obvious way to mock around it (though there are likely long/complicated ways to do so).

GeoServer is a very old application (over 15 years), working with it calls for practicality rather than perfectionism.

Reply:
ok, I'll leave it where it was.